File: DBMS-2/apriori.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int main() {
    // Read CSV file
    vector<vector<string>> data;
    ifstream file("data.csv");
    string line;
    while (getline(file, line)) {
        vector<string> elements;
        string element;
        for (char c : line) {
            if (c == ',') {
                elements.push_back(element);
                element.clear();
            } else {
                element += c;
            }
        }
        elements.push_back(element);
        data.push_back(elements);
    }

    // Set minimum support threshold
    int min_support = 2;

    // Perform Apriori algorithm
    vector<vector<string>> frequent_sets;
    for (int k = 1; k <= data[0].size(); k++) {
        // Generate candidate item sets
        vector<vector<string>> candidates;
        if (k == 1) {
            for (string item : data[0]) {
                candidates.push_back({item});
            }
        } else {
            for (vector<string> set1 : frequent_sets) {
                for (vector<string> set2 : frequent_sets) {
                    if (set1[0] < set2[0]) {
                        vector<string> candidate = set1;
                        candidate.push_back(set2[0]);
                        candidates.push_back(candidate);
                    }
                }
            }
        }

        // Check support for candidate item sets
        vector<vector<string>> frequent;
        for (vector<string> candidate : candidates) {
            int count = 0;
            for (vector<string> row : data) {
                bool contains = true;
                for (string item : candidate) {
                    if (find(row.begin(), row.end(), item) == row.end()) {
                        contains = false;
                        break;
                    }
                }
                if (contains) {
                    count++;
                }
            }
            if (count >= min_support) {
                frequent.push_back(candidate);
            }
        }

        // Update frequent item sets
        frequent_sets = frequent;

        // Print frequent item sets
        for (vector<string> set : frequent_sets) {
            for (string item : set) {
                cout << item << " ";
            }
            cout << endl;
        }
    }

      // Generate association rules
    for (vector<string> set : frequent_sets) {
        for (int i = 0; i < set.size(); i++) {
            vector<string> antecedent(set.begin(), set.begin() + i);
            vector<string> consequent(set.begin() + i, set.end());
            if (!antecedent.empty() && !consequent.empty()) {
                int support = 0;
                for (vector<string> row : data) {
                    bool contains = true;
                    for (string item : set) {
                        if (find(row.begin(), row.end(), item) == row.end()) {
                            contains = false;
                            break;
                        }
                    }
                    if (contains) {
                        support++;
                    }
                }

// Count the number of transactions that contain the antecedent and consequent
// int support = 0;
for (vector<string> transaction : data) {
    bool contains_antecedent = true;
    bool contains_consequent = false;
    for (string item : antecedent) {
        if (find(transaction.begin(), transaction.end(), item) == transaction.end()) {
            contains_antecedent = false;
            break;
        }
    }
    for (string item : consequent) {
        if (find(transaction.begin(), transaction.end(), item) != transaction.end()) {
            contains_consequent = true;
            break;
        }
    }
    if (contains_antecedent && contains_consequent) {
        support++;
    }
}


// Calculate confidence
// double confidence = (double)support / antecedent_support;

                double confidence = ((double)support / count(data.begin(), data.end(), antecedent))*100;
                cout << "Rule: ";
                for (string item : antecedent) {
                    cout << item << " ";
                }
                cout << "-> ";
                for (string item : consequent) {
                    cout << item << " ";
                }
                cout << " (support = " << support << ", confidence = " << confidence << ")" << endl;
            }
        }
    }
}

File: DBMS-2/apriori.py
import csv

# A function to return the frequent item sets with a minimum support
def apriori(dataset, min_support):
    C1 = {}
    for transaction in dataset:
        for item in transaction:
            if item in C1:
                C1[item] += 1
            else:
                C1[item] = 1
    
    # Remove items with support less than min_support
    C1 = {k:v for k,v in C1.items() if v >= min_support}
    L1 = list(C1.keys())

    L = [L1]
    k = 2
    while (len(L[k-2]) > 0):
        Ck = {}
        for i in range(len(L[k-2])):
            for j in range(i+1, len(L[k-2])):
                a = list(L[k-2][i])
                a.sort()
                b = list(L[k-2][j])
                b.sort()
                if a[0:k-2] == b[0:k-2]:
                    c = a + [b[k-2]]
                    c.sort()
                    Ck[frozenset(c)] = 0
        for transaction in dataset:
            for item in Ck:
                if item.issubset(transaction):
                    Ck[item] += 1

        Lk = set(item for item in Ck if Ck[item] >= min_support)
        L.append(Lk)
        k += 1
    return L

# Read data from CSV file
transactions = []
with open('data.csv') as csvfile:
    reader = csv.reader(csvfile,delimiter=",")
    for row in reader:
        transactions.append(row)
# Calculate the confidence of the association rules
frequent_item_sets = apriori(transactions, min_support=2)
print(frequent_item_sets)
confidence = {}
for k in range(1, len(frequent_item_sets)):
    for itemset in frequent_item_sets[k]:
        for item in itemset:
            antecedent = frozenset([item])
            consequent = itemset - antecedent
            if antecedent in confidence:
                confidence[antecedent][consequent] = frequent_item_sets[k][itemset] / frequent_item_sets[k-1][antecedent]
            else:
                confidence[antecedent] = {consequent: frequent_item_sets[k][itemset] / frequent_item_sets[k-1][antecedent]}

# Print the confidence of the association rules
            for antecedent in confidence:
                for consequent in confidence[antecedent]:
                    # print("(((((((((((((((((")
                    print("Rule: {} -> {}  Confidence: {}".format(antecedent, consequent, confidence[antecedent][consequent]))

# Run Apriori algorithm
print("JJJJJJJJJJJJJJJJJJJJ")